pkg/api/v1/atlascluster_types.go

```diff
@@ -17,23 +17,231 @@ limitations under the License.
+	InstanceSizeName string `json:"instanceSizeName,omitempty"`
+
+	// Cloud service provider on which Atlas provisions the hosts.
+	// +kubebuilder:validation:Enum=AWS;GCP;AZURE;TENANT
+	ProviderName string `json:"providerName,omitempty"`
+
+	// Physical location of your MongoDB cluster.
+	// The region you choose can affect network latency for clients accessing your databases.
+	// +optional
+	RegionName string `json:"regionName,omitempty"`
+
+	// Disk IOPS setting for AWS storage.
+	// Set only if you selected AWS as your cloud service provider.
+	// +kubebuilder:validation:Enum=STANDARD;PROVISIONED
+	VolumeType string `json:"volumeType,omitempty"`
+
+	// Range of instance sizes to which your cluster can scale.
+	AutoScaling *AutoScalingSpec `json:"autoScaling,omitempty"`
+}
+
+// ReplicationSpec represents a configuration for cluster regions
+type ReplicationSpec struct {
+	// Number of shards to deploy in each specified zone.
+	// The default value is 1.
+	NumShards *int64 `json:"numShards,omitempty"`
+
+	// Name for the zone in a Global Cluster.
+	// Don't provide this value if clusterType is not GEOSHARDED.
+	// +optional
+	ZoneName string `json:"zoneName,omitempty"`
+
+	// Configuration for a region.
+	// Each regionsConfig object describes the region's priority in elections and the number and type of MongoDB nodes that Atlas deploys to the region.
+	// +optional
+	RegionsConfig map[string]RegionsConfig `json:"regionsConfig,omitempty"`
+}
+
+// RegionsConfig describes the region’s priority in elections and the number and type of MongoDB nodes Atlas deploys to the region.
+type RegionsConfig struct {
+	// The number of analytics nodes for Atlas to deploy to the region.
+	// Analytics nodes are useful for handling analytic data such as reporting queries from BI Connector for Atlas.
+	// Analytics nodes are read-only, and can never become the primary.
+	// If you do not specify this option, no analytics nodes are deployed to the region.
+	// +optional
+	AnalyticsNodes *int64 `json:"analyticsNodes,omitempty"`
+
+	// Number of electable nodes for Atlas to deploy to the region.
+	// Electable nodes can become the primary and can facilitate local reads.
+	// +optional
+	ElectableNodes *int64 `json:"electableNodes,omitempty"`
+
+	// Election priority of the region.
+	// For regions with only replicationSpecs[n].regionsConfig.<region>.readOnlyNodes, set this value to 0.
+	// +optional
+	Priority *int64 `json:"priority,omitempty"`
+
+	// Number of read-only nodes for Atlas to deploy to the region.
+	// Read-only nodes can never become the primary, but can facilitate local-reads.
+	// +optional
+	ReadOnlyNodes *int64 `json:"readOnlyNodes,omitempty"`
+}
+
+// Cluster converts the Spec to native Atlas client format.
+func (spec *AtlasClusterSpec) Cluster() *mongodbatlas.Cluster {
+	result := &mongodbatlas.Cluster{}
+	err := copier.Copy(result, spec)
+	if err != nil {
+		panic(err)
```

Comment:
we need to handle these kinds of exceptions

Reply:
this panic would mean we have incompatibilities in our types, which is not something the user can fix without altering the code - so how should we handle this?

Reply:
besides, from my observations, we don't handle similar panics in other PRs/existing code - should we?

Reply:
One thing: in case of panic the whole operator falls and will fall on retry as well (as the event will come again) so it will get stuck. 
Catching this bug will isolate the problem for one single resource and will result in a proper status message that the customer will read and maybe will manage to fix the problem by removing some fields.
I agree that it's possible to panic on some internal development logic that doesn't depend on user input - so the tests would catch it. But here the `cluster` is the user-provided spec and they may change it to fix the problem

Reply:
I agree we should be able to gracefully handle incorrect input. This logic, however, does _not_ handle any user input, as it only converts between two different statically-typed `struct`s, and it is our job to make sure they are correct and convertible.

Even after I change this to JSON S/D, any error during (de)serialization will mean something is broken in our code beyond repair, and I think we should fail as loud as possible whenever that's the case - hence the panic.

Am I missing something here?

Reply:
> This logic, however, does not handle any user input
Not sure I agree with this as the logic deals with the AtlasCluster `spec` which is provided by the user. The Spec is taken and transformed into the atlas-client struct

Reply:
As discussed 1:1, the panics were removed